### PR TITLE
create a dummy card when creating new study

### PR DESCRIPTION
**Pull request by JacquesBorsenberger**
create a dummy card when creating new study
Signed-off-by: jacqbors <jacques.borsenberger@rte-france.com>

### CODE REVIEW

File: src/components/create-study-form.js [hunk 1]
```diff
@@ -268,12 +268,14 @@ export const CreateStudyForm = () => {
...
-            isPrivateStudy = true;
-        }
+        let isPrivateStudy = studyPrivacy === 'private';
 
         setOpen(false);
+        props.addCreationRequest({
+            studyName: studyName,
+            userId: userId,
+            isPrivate: isPrivateStudy,
+        });
```
Comment (niconoir) [inline src/components/create-study-form.js:278]: Parameter "date" is not set when calling addCreationRequest, is it normal ? When the temporary study card is displayed, an "Invalid date" label is then displayed.

**Review by niconoir — CHANGES_REQUESTED** [review]

**Review by niconoir — APPROVED** [review]

File: src/components/study-manager.js [hunk 2]
```diff
@@ -119,6 +119,10 @@ const CustomTypography = withStyles({
     },
 })(Typography);
 
+function makeKey({ userId, studyName }) {
```
Comment (SlimaneAmar) [inline src/components/study-manager.js:122]: Why a key with the userId ?

File: src/components/study-manager.js [hunk 3]
```diff
@@ -551,16 +557,71 @@ const StudyManager = ({ onClick }) => {
...
+
+    const cleanLocalCreationRequest = useCallback(
+        (remote) => {
+            if (localCreationRequest) {
+                let didDelete = false;
+                remote.forEach((study) => {
+                    if (localCreationRequest.hasOwnProperty(makeKey(study))) {
+                        didDelete = true;
+                        delete localCreationRequest[makeKey(study)];
+                    }
```
Comment (SlimaneAmar) [inline src/components/study-manager.js:581]: Maybe it's necessary to do a copy of the state before modifying it

File: src/components/study-manager.js [hunk 4]
```diff
@@ -551,16 +557,71 @@ const StudyManager = ({ onClick }) => {
...
+
+    useEffect(() => {
+        cleanLocalCreationRequest(studyCreationRequests);
+    }, [studyCreationRequests, cleanLocalCreationRequest]);
+
+    useEffect(() => {
+        cleanLocalCreationRequest(studies);
+    }, [studies, cleanLocalCreationRequest]);
+
+    function mergeCreationRequest(remote, local) {
```
Comment (SlimaneAmar) [inline src/components/study-manager.js:597]: mergeCreationRequests ?

File: src/components/study-manager.js [hunk 5]
```diff
@@ -551,16 +557,71 @@ const StudyManager = ({ onClick }) => {
...
+    function addCreationRequest({ studyName, userId, ...rest }) {
+        setLocalCreationRequest({
+            ...localCreationRequest,
+            ...{
+                [makeKey({ userId: userId, studyName: studyName })]: {
+                    studyName: studyName,
+                    userId: userId,
+                    ...rest,
+                },
+            },
```
Comment (SlimaneAmar) [inline src/components/study-manager.js:569]: Maybe use a map to improve readability
```
        setLocalCreationRequest(
            new Map(localCreationRequest.entries()).set(makeKey(studyInfos), studyInfos)
        );
```

**Review by SlimaneAmar — APPROVED** [review]

File: src/components/study-manager.js [hunk 6]
```diff
@@ -551,16 +557,71 @@ const StudyManager = ({ onClick }) => {
...
+        (remote) => {
+            if (localCreationRequests) {
+                let didDelete = false;
+                remote.forEach((study) => {
+                    if (localCreationRequests.hasOwnProperty(makeKey(study))) {
+                        didDelete = true;
+                        delete localCreationRequests[makeKey(study)];
+                    }
+                });
+                if (didDelete) setlocalCreationRequests(localCreationRequests);
```
Comment (jonenst) [inline src/components/study-manager.js:583]: Does this do anything ? React uses shallow equality on arrays to check so it should be a noop? Maybe that's why we see the duplicate study for a split second ?

**Comment by jonenst** [comment]
when there is an error with the study creation, the study creation requests is displayed indefinitely until the page is refreshed.

File: src/components/study-manager.js (hunk 6 above)
Comment (JacquesBorsenberger) [reply src/components/study-manager.js]: changed

File: src/components/study-manager.js [hunk 7]
```diff
@@ -485,6 +489,9 @@ const StudyManager = ({ onClick }) => {
 
     const classes = useStyles();
 
+    const [localCreationRequests, setlocalCreationRequests] = useState({});
+    const sutdyCreationSubmitted = useRef(new Set());
```
Comment (jonenst) [inline src/components/study-manager.js:493]: typo sutdy -> study

**Review by jonenst — APPROVED** [review]

**Pull request merged**